zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -769,6 +811,164 @@ impl<DbT: InputSource> InputSelector for GreedyInputSelector<DbT> {
+                0,
+                spends_orchard,
+            ),
+            Address::Unified(addr) => {
+                if cfg!(feature = "orchard") && addr.has_orchard() {
+                    change_strategy.fee_rule().fee_required(
+                        params,
+                        target_height,
+                        [],
+                        [],
+                        spends_sapling,
+                        0,
+                        std::cmp::max(spends_orchard, 1),
+                    )
+                } else if addr.has_sapling() {
+                    change_strategy.fee_rule().fee_required(
+                        params,
+                        target_height,
+                        [],
+                        [],
+                        spends_sapling,
+                        1,
+                        spends_orchard,
+                    )
+                } else if addr.has_transparent() {
+                    change_strategy.fee_rule().fee_required(
+                        params,
+                        target_height,
+                        [],
+                        [P2PKH_STANDARD_OUTPUT_SIZE],
+                        spends_sapling,
+                        0,
+                        spends_orchard,
+                    )
+                } else {
+                    unreachable!()
+                }
+            }
+            Address::Tex(_) => change_strategy
+                .fee_rule()
+                .fee_required(
+                    params,
+                    target_height,
+                    [],
+                    [P2PKH_STANDARD_OUTPUT_SIZE],
+                    spendable_notes.sapling().len(),
+                    match spends_sapling > 0 && spends_orchard == 0 {
+                        true => 2,
+                        false => 0,
+                    },
+                    if spends_orchard > 0 {
+                        std::cmp::max(spends_orchard, 2)
+                    } else {
+                        0
+                    },
+                )
+                .and_then(|t0_fee| {
+                    let t1_fee = change_strategy.fee_rule().fee_required(
+                        params,
+                        target_height,
+                        [InputSize::Known(P2PKH_STANDARD_INPUT_SIZE)],
+                        [P2PKH_STANDARD_OUTPUT_SIZE],
+                        0,
+                        0,
+                        0,
+                    )?;
+
+                    Ok((t0_fee + t1_fee).expect("fee is in range of valid Zatoshis values"))
+                }),
+        }
+        .map_err(|fee_error| {
+            InputSelectorError::Change(ChangeError::StrategyError(ChangeT::Error::from(fee_error)))
+        })?;
+
+        let amount = (input_total - fee_required).ok_or(BalanceError::Underflow)?;
+
+        let payment = match zip321::Payment::new(recipient, amount, memo, None, None, vec![]) {
```

Comment:
Firse, use `ok_or_else` instead of this `match`. Second, please open an issue to make `Payment::new` return a `Result` with an informative error type instead of just an `Option`.

Reply:
Created the issue https://github.com/zcash/librustzcash/issues/1794